src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -46,23 +79,69 @@ private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
                     _openDocumentContexts.Remove(e.Document.Id);
 
                     // Ensure we dispose the context, which we'll do asynchronously
-                    contextTask.ContinueWith(
-                        t => t.Result.Dispose(),
-                        CancellationToken.None,
-                        TaskContinuationOptions.OnlyOnRanToCompletion,
-                        TaskScheduler.Default);
+                    EnsureContextCleanup(contextTask);
                 }
             }
         }
 
-        private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
+        private void OnDocumentOpened(object sender, DocumentEventArgs e)
         {
             lock (_gate)
             {
-                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+                var contextTask = Task.Run(async () =>
+                {
+                    var context = await GetConventionContextAsync(e.Document.FilePath, CancellationToken.None).ConfigureAwait(false);
+                    context.CodingConventionsChangedAsync += OnCodingConventionsChangedAsync;
+                    return context;
+                });
+
+                Contract.Requires(!_openDocumentContexts.ContainsKey(e.Document.Id));
+                _openDocumentContexts.Add(e.Document.Id, contextTask);
             }
         }
 
+        private void ClearOpenFileCache(ProjectId projectId = null)
+        {
+            var contextTasks = new List<Task<ICodingConventionContext>>();
+
+            lock (_gate)
+            {
+                if (projectId == null)
+                {
+                    contextTasks.AddRange(_openDocumentContexts.Values);
+                    _openDocumentContexts.Clear();
+                }
+                else
+                {
+                    foreach (var kv in _openDocumentContexts.Where(kv => kv.Key.ProjectId == projectId).ToList())
+                    {
+                        _openDocumentContexts.Remove(kv.Key);
+                        contextTasks.Add(kv.Value);
+                    }
+                }
+            }
+
+            foreach (var contextTask in contextTasks)
+            {
+                EnsureContextCleanup(contextTask);
+            }
+        }
+
+        private void EnsureContextCleanup(Task<ICodingConventionContext> contextTask)
+        {
+            contextTask.ContinueWith(
+                t =>
+                {
+                    var context = t.Result;
+
+                    context.CodingConventionsChangedAsync -= OnCodingConventionsChangedAsync;
+                    context.Dispose();
+                },
+                CancellationToken.None,
+                TaskContinuationOptions.OnlyOnRanToCompletion,
```

Comment:
this worries me.  this means we may not dispose hte context in all cases.  Note that that's not usually the contract of "try/finally" which will run the dispose even if an exception is thrown.

Note: this would not be an issue if this was an async/await function and we just used normal flow control constructs.  I would highly recommend that instead of manully writing explicit ContinueWith code for features that do not need it.

Reply:
@jasonmalinowski your call. if you want to, we can change to different pattern.